pageserver/src/tenant/timeline.rs

```diff
@@ -3006,34 +2997,49 @@ impl Timeline {
     }
 
     // Write out the given frozen in-memory layer as a new L0 delta file
-    fn create_delta_layer(
+    async fn create_delta_layer(
         self: &Arc<Self>,
         frozen_layer: &InMemoryLayer,
     ) -> anyhow::Result<(LayerFileName, LayerFileMetadata)> {
-        // Write it out
-        let new_delta = frozen_layer.write_to_disk()?;
-        let new_delta_path = new_delta.path();
-        let new_delta_filename = new_delta.filename();
+        // TODO figure out how to use spawn_blocking. Can't use it because frozen_layer is not 'static
 
-        // Sync it to disk.
-        //
-        // We must also fsync the timeline dir to ensure the directory entries for
-        // new layer files are durable
-        //
-        // TODO: If we're running inside 'flush_frozen_layers' and there are multiple
-        // files to flush, it might be better to first write them all, and then fsync
-        // them all in parallel.
+        let span = tracing::info_span!("blocking");
+        let (new_delta, sz): (DeltaLayer, _) = tokio::task::block_in_place({
+            let _g = span.entered();
+            let self_clone = Arc::clone(self);
+            move || {
+                // Write it out
+                let new_delta = frozen_layer.write_to_disk()?;
+                let new_delta_path = new_delta.path();
 
-        // First sync the delta layer. We still use par_fsync here to keep everything consistent. Feel free to replace
-        // this with a single fsync in future refactors.
-        par_fsync::par_fsync(&[new_delta_path.clone()]).context("fsync of delta layer")?;
-        // Then sync the parent directory.
-        par_fsync::par_fsync(&[self.conf.timeline_path(&self.timeline_id, &self.tenant_id)])
-            .context("fsync of timeline dir")?;
+                // Sync it to disk.
+                //
+                // We must also fsync the timeline dir to ensure the directory entries for
+                // new layer files are durable
+                //
+                // TODO: If we're running inside 'flush_frozen_layers' and there are multiple
+                // files to flush, it might be better to first write them all, and then fsync
+                // them all in parallel.
+
+                // First sync the delta layer. We still use par_fsync here to keep everything consistent. Feel free to replace
+                // this with a single fsync in future refactors.
+                par_fsync::par_fsync(&[new_delta_path.clone()]).context("fsync of delta layer")?;
+                // Then sync the parent directory.
+                par_fsync::par_fsync(&[self_clone
+                    .conf
+                    .timeline_path(&self_clone.timeline_id, &self_clone.tenant_id)])
+                .context("fsync of timeline dir")?;
```

Comment:
we can use `par_fsync_async` if we move this outside the task_in_place block.

Reply:
Not sure if that's actually better in terms of performance, because, these async variants internally just use spawn_blocking as well.
So we'd just add an additional spawn_blocking call worth of latency.

Now I don't know if the same call can be made for block_in_place. AFAIK it just informs tokio runtime that the current worker thread might block?
So, maybe it's better to keep the block_in_place as short as possible?
In that case, your suggestion makes sense.

Would appreciate comment from someone with more tokio experience

Reply:
I would prefer async version because it uses shared tokio threadpool instead of our own that spawns threads for each call.

Reply:
But we need the spawn_blocking because the writing of the delta layer directly above this piece of code is currently using sync code...

Reply:
Hm, dont see any problems there, spawn_blocking to write then await on sync that uses spawn blocking as well. It will reuse threads from tokio pool. Though an important detail here is that there is no actual parallelism happening because with `paths.len() == 1` it wont spawn additional threads. So maybe we can just fsync them without `par` for clarity? Then it will be in the same spawn_blocking until `write_to_disk` becomes async, wdyt?